src/librustc_metadata/decoder.rs

```diff
@@ -1148,12 +1148,6 @@ impl<'a, 'tcx> CrateMetadata {
 
         // Lock the codemap_import_info to ensure this only happens once
         let mut codemap_import_info = self.codemap_import_info.borrow_mut();
-
-        if !codemap_import_info.is_empty() {
```

Comment:
The check is not really unnecessary, I think. The `codemap_import_info` field is a read-write-lock, so we have to check that it has not been filled concurrently by another thread. You could add a comment to that effect.

Reply:
@michaelwoerister there is an identical check right above this (except avoiding the mutable borrow); one of them should be redundant, right?

Reply:
Consider the following:
```
thread 1: acquire read lock 
thread 1: check is_empty() -> true 
thread 1: release read lock
thread 2: acquire read lock
thread 2: check is_empty() -> true
thread 2: release read lock
thread 2: acquire write lock
thread 1: acquire write lock -> blocked
thread 2: initialize locked data
thread 2: release write lock
thread 1: unblocked -> acquire write lock
thread 1: initialize data - although it was already initialized by thread 2

```

Reply:
Ooo, I'm with you now. That's a tricky one :)

Reply:
It's also not ideal that this is hard to even identify as a multi-threaded context.